core/base/v7/inc/ROOT/RError.hxx

```diff
@@ -0,0 +1,230 @@
+   std::vector<RLocation> fStackTrace;
+
+public:
+   /// Used by R__FAIL
+   RError(const std::string &message, const std::string &func, const std::string &file, int line);
+   /// Used by R__FORWARD_RESULT
+   void AddFrame(const std::string &func, const std::string &file, int line);
+   /// Add more information to the diagnostics
+   void AppendMessage(const std::string &info) { fMessage += info; }
+   /// Format a dignostics report, e.g. for an exception message
+   std::string GetReport() const;
+   const std::vector<RLocation> &GetStackTrace() const { return fStackTrace; }
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::RException
+\ingroup Base
+\brief Base class for all ROOT issued exceptions
+*/
+// clang-format on
+class RException : public std::runtime_error {
+   RError fError;
+public:
+   explicit RException(const RError &error) : std::runtime_error(error.GetReport()), fError(error) {}
+   const RError &GetError() const { return fError; }
+};
+
+
+/// Wrapper class that generates a data member of type T in RResult<T> for all Ts except T == void
+namespace Internal {
+template <typename T>
+struct RResultType {
+   RResultType() = default;
+   explicit RResultType(const T &value) : fValue(value) {}
+   T fValue;
+};
+
+class RResultTypeVoid {};
+} // namespace Internal
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::RResult
+\ingroup Base
+\brief The class is used as a return type for operations that can fail; wraps a value of type T or an RError
+
+The RResult enforces checking whether it contains a valid value or an error state. If the RResult leaves the scope
+unchecked, it will throw an exception.  RResult should only be allocated on the stack, which is helped by deleting the
+new operator.  RResult is movable but not copyable to avoid throwing exceptions multiple times.
+*/
+// clang-format on
+template <typename T>
+class RResult : public std::conditional_t<std::is_void<T>::value, Internal::RResultTypeVoid, Internal::RResultType<T>> {
+private:
+   /// This is the nullptr for an RResult representing success
+   std::unique_ptr<RError> fError;
+   /// This is safe because checking an RResult is not a multi-threaded operation
+   mutable bool fIsChecked{false};
+
+public:
+   /// Constructor is _not_ explicit in order to allow for `return T();` for functions returning RResult<T>
+   /// Only available if T is not void
+   template <typename Dummy = T, typename = typename std::enable_if_t<!std::is_void<T>::value, Dummy>>
+   RResult(const Dummy &value) : Internal::RResultType<T>(value) {}
+   /// Constructor is _not_ explicit such that the RError returned by R__FAIL can be converted into an RResult<T>
+   /// for any T
+   RResult(const RError &error) : fError(std::make_unique<RError>(error)) {}
+
+   /// RResult<void> has a default contructor that creates an object representing success
+   template <typename Dummy = T, typename = typename std::enable_if_t<std::is_void<T>::value, Dummy>>
+   RResult() {}
```

Comment:
Instead of a `static const RResult` there is now a `static RResult Success()` for `T == void`, so that users can do `return RResult<void>::Success()` (instead of `R__SUCCESS`). Returning a `static const RResult kVoidSuccess` would have involved a copy and `RResult` is not copyable.